### PR TITLE
Calculate AWS Signature for file body type

### PR DESCRIPTION
**Pull request by vikiCoder**
Related to issue [#3232](​https://github.com/postmanlabs/postman-app-support/issues/3232) - AWS SIGV4 signature mismatch with "binary" payload

Potentially resolves following issues:
https://github.com/postmanlabs/postman-app-support/issues/3232
https://github.com/postmanlabs/postman-app-support/issues/4127
​https://github.com/postmanlabs/postman-app-support/issues/5996
https://github.com/postmanlabs/postman-app-support/issues/4151
https://github.com/postmanlabs/postman-app-support/issues/5873

Todo:
- [x] Signature calculation for body type: file
- [ ] Signature calculation for body type: form-data
- [x] Signature calculation for body type: urlencoded

### CODE REVIEW

File: lib/authorizer/aws4.js [hunk 1]
```diff
@@ -92,6 +97,46 @@ module.exports = {
...
+        }
+
+        if (body.mode === RequestBody.MODES.formdata) {
+            // @todo: This requires adding new data to form instead of setting headers for AWS auth
+
+            return callback();
+        }
+
+        if (body.mode === RequestBody.MODES.file) {
+            util.createReadStream(fileResolver, body.file.src, function (err, stream) {
```
Comment (shamasis) [inline lib/authorizer/aws4.js]: three level access needs a check.

File: lib/authorizer/auth-interface.js [hunk 2]
```diff
@@ -10,12 +10,14 @@ var _ = require('lodash'),
...
  * @throws {Error}
  */
-createAuthInterface = function (auth) {
+createAuthInterface = function (auth, fileResolver) {
     if (!(auth && auth.parameters && auth.parameters())) {
         throw new Error('runtime~createAuthInterface: invalid auth');
     }
 
     return /** @lends AuthInterface.prototype **/{
+        fileResolver: fileResolver,
```
Comment (shamasis) [inline lib/authorizer/auth-interface.js]: security risk

File: lib/authorizer/aws4.js [hunk 3]
```diff
@@ -92,6 +97,75 @@ module.exports = {
...
+     *
+     * @todo This function can also be used in Digest auth so that it works correctly for urlencoded and file body types
+     *
+     * @param {RequestBody} body
+     * @param {String} algorithm
+     * @param {String} digestEncoding
+     * @param {Function} callback
+     */
+    computeBodyHash: function (body, algorithm, digestEncoding, callback) {
+        if (!body || !algorithm || !digestEncoding) { return callback(); }
```
Comment (codenirvana) [inline lib/authorizer/aws4.js]:
```suggestion
        if (!(body && algorithm && digestEncoding)) { return callback(); }
```

File: lib/authorizer/aws4.js [hunk 4]
```diff
@@ -92,6 +97,75 @@ module.exports = {
...
+     * Calculates body hash with given algorithm and digestEncoding.
+     *
+     * @todo This function can also be used in Digest auth so that it works correctly for urlencoded and file body types
+     *
+     * @param {RequestBody} body
+     * @param {String} algorithm
+     * @param {String} digestEncoding
+     * @param {Function} callback
+     */
+    computeBodyHash: function (body, algorithm, digestEncoding, callback) {
```
Comment (codenirvana) [inline lib/authorizer/aws4.js]: Don't export this, convert to a private method.

File: lib/authorizer/aws4.js [hunk 5]
```diff
@@ -92,6 +97,75 @@ module.exports = {
...
+        }
+    },
+
+    /**
+     * Encodes characters not encoded by querystring.stringify() according to RFC3986.
+     *
+     * @param {String} str Partially encoded string by querystring.stringify()
+     * @returns {String} Fully encoded string
+     */
+    rfc3986: function (str) {
```
Comment (codenirvana) [inline lib/authorizer/aws4.js]: Don't export this, convert to a private method.
Also add its origin/reference.

File: package.json [hunk 6]
```diff
@@ -43,6 +43,7 @@
     "postman-collection": "3.4.6-beta.1",
     "postman-request": "2.88.1-postman.7.1",
     "postman-sandbox": "3.2.5",
+    "qs": "6.6.0",
```
Comment (codenirvana) [inline package.json]: this is not required

**Pull request merged**